specification/dartLangSpec.tex

```diff
@@ -2073,30 +2073,38 @@ \section{Functions}
 a function marked \code{\SYNC*} or \code{\ASYNC*} is \VOID.
 
 \LMHash{}%
-We define the notion of the
-\IndexCustom{element type of a generator}{function!generator!element type}
-as follows:
+We define the
+\Index{union-free type}
+of a type $T$ as follows:
+If $T$ is of the form \code{$S$?}\ or the form \code{FutureOr<$S$>}
+then the union-free type of $T$ is the union-free type of $S$.
+Otherwise, the union-free type of $T$ is $T$.
+\commentary{%
+For example, the union-free type of \code{FutureOr<int?>?} is \code{int}.%
+}
+
+\LMHash{}%
+We define the
+\IndexCustom{element type of a generator function}{%
+    function!generator!element type}
+$f$ as follows:
+%
+Let $S$ be the union-free type of the declared return type of $f$.
```

Comment:
Can we just reject programs where the declared return type is not either `Iterable` (`Stream` respectively) or a top type?

Reply:
If think this might reject some valid programs. It's a long shot ("come on, who would write a program like that?"), but the following example could actually occur in some context where the developer doesn't have much of a choice (example from [here](https://github.com/dart-lang/language/issues/3148#issuecomment-1649253527)):

```dart
class A {
  FutureOr<Iterable<int>?> m() => null;
}

class B extends A {
  FutureOr<Iterable<int>?> m() sync* { yield 1; yield 2; }
}

class C extends B {
  FutureOr<Iterable<int>?> m() => Future.value();
}
```

Reply:
Ack.